pyop2/mpi.py

```diff
@@ -33,9 +33,184 @@
+# on any dupped communicators (rather than relying on the garbage collector).
+#
+# Finally, since it's difficult to know when all these communicators
+# go out of scope, we register an atexit handler to clean up any
+# outstanding duplicated communicators.
+
+
+def delcomm_outer(comm, keyval, icomm):
+    """Deleter for internal communicator, removes reference to outer comm.
+
+    :arg comm: Outer communicator.
+    :arg keyval: The MPI keyval, should be ``innercomm_keyval``.
+    :arg icomm: The inner communicator, should have a reference to
+        ``comm`.
+    """
+    if keyval != innercomm_keyval:
+        raise ValueError("Unexpected keyval")
+    ocomm = icomm.Get_attr(outercomm_keyval)
+    if ocomm is None:
+        raise ValueError("Inner comm does not have expected reference to outer comm")
+
+    if ocomm != comm:
+        raise ValueError("Inner comm has reference to non-matching outer comm")
+    icomm.Delete_attr(outercomm_keyval)
+
+
+# Refcount attribute for internal communicators
+refcount_keyval = MPI.Comm.Create_keyval()
+
+# Inner communicator attribute (attaches inner comm to user communicator)
+innercomm_keyval = MPI.Comm.Create_keyval(delete_fn=delcomm_outer)
+
+# Outer communicator attribute (attaches user comm to inner communicator)
+outercomm_keyval = MPI.Comm.Create_keyval()
+
+# List of internal communicators, must be freed at exit.
+dupped_comms = []
+
+
+def dup_comm(comm_in=None):
+    """Given a communicator return a communicator for internal use.
+
+    :arg comm_in: Communicator to duplicate.  If not provided,
+        defaults to COMM_WORLD.
+
+    :returns: An mpi4py communicator."""
+    if comm_in is None:
+        comm_in = COMM_WORLD
+    if isinstance(comm_in, PETSc.Comm):
+        comm_in = comm_in.tompi4py()
+    elif not isinstance(comm_in, MPI.Comm):
+        raise ValueError("Don't know how to dup a %r" % type(comm_in))
+    if comm_in == MPI.COMM_NULL:
+        return comm_in
+    refcount = comm_in.Get_attr(refcount_keyval)
+    if refcount is not None:
+        # Passed an existing PyOP2 comm, return it
+        comm_out = comm_in
+        refcount[0] += 1
+    else:
+        # Check if communicator has an embedded PyOP2 comm.
+        comm_out = comm_in.Get_attr(innercomm_keyval)
+        if comm_out is None:
+            # Haven't seen this comm before, duplicate it.
+            comm_out = comm_in.Dup()
+            comm_in.Set_attr(innercomm_keyval, comm_out)
+            comm_out.Set_attr(outercomm_keyval, comm_in)
+            # Refcount
+            comm_out.Set_attr(refcount_keyval, [1])
+            # Remember we need to destroy it.
+            dupped_comms.append(comm_out)
+        else:
+            refcount = comm_out.Get_attr(refcount_keyval)
+            if refcount is None:
+                raise ValueError("Inner comm without a refcount")
+            refcount[0] += 1
+    return comm_out
+
+
+def free_comm(comm, remove=True):
```

Comment:
Ok, I see, so `dup_comm` and `free_comm` appear in pairs like with a context manager, and it provides "lexical scoping" for the communicator.